### PR TITLE
=str Add IterableSource.

### PR DESCRIPTION
**Pull request by He-Pin**
was https://github.com/akka/akka/pull/31372

This introduce a dedicated IterableSource, which can be later used for performance optimization. And also obey the supervisionStrategy now.

### CODE REVIEW

File: stream-tests/src/test/scala/org/apache/pekko/stream/scaladsl/SourceSpec.scala [hunk 1]
```diff
@@ -311,19 +311,60 @@ class SourceSpec extends StreamSpec with DefaultTimeout {
...
+        .futureValue should ===(List(1, 2, 1, 2, 1, 2, 1, 2, 1, 2))
 
       Source
         .fromIterator(() => (1 to 5).toIterator.map(_ => throw TE("b")))
         .withAttributes(ActorAttributes.supervisionStrategy(Supervision.restartingDecider))
         .grouped(10)
         .runWith(Sink.headOption)
-        .futureValue should ===(None)
+        .failed
+        .futureValue shouldBe an[TE]
```
Comment (pjfanning) [inline stream-tests/src/test/scala/org/apache/pekko/stream/scaladsl/SourceSpec.scala:351]: `shouldBe a[TE]` would be better

**Review by pjfanning — APPROVED** [review]
lgtm - but I'd prefer if some other reviewers also had a look before we merge

File: stream-tests/src/test/scala/org/apache/pekko/stream/scaladsl/SourceSpec.scala [hunk 2]
```diff
@@ -311,19 +311,60 @@ class SourceSpec extends StreamSpec with DefaultTimeout {
     }
 
     "use decider when iterator throws" in {
+
```
Comment (nvollmar) [inline stream-tests/src/test/scala/org/apache/pekko/stream/scaladsl/SourceSpec.scala:314]: I'd prefer to have separate tests with descriptive titles if they test different edge cases, or at least a comment per run hinting what each is testing.

File: stream/src/main/scala/org/apache/pekko/stream/impl/fusing/IterableSource.scala [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+                tryPushNextOrComplete()
+            }
+        }
+
+      private def tryPushNextOrComplete(): Unit =
+        if (currentIterator.hasNext) {
+          if (isAvailable(out)) {
+            push(out, currentIterator.next())
+          }
+          if (!currentIterator.hasNext) {
```
Comment (nvollmar) [inline stream/src/main/scala/org/apache/pekko/stream/impl/fusing/IterableSource.scala:65]: This check only makes sense if we pushed an element -> move it inside the `if (isAvailable(out)) {` block

File: stream/src/main/scala/org/apache/pekko/stream/impl/fusing/IterableSource.scala [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+  override val shape: SourceShape[T] = SourceShape(out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new GraphStageLogic(shape) with OutHandler {
+      private lazy val decider = inheritedAttributes.mandatoryAttribute[SupervisionStrategy].decider
+      private var currentIterator: Iterator[T] = _
+
+      override def onPull(): Unit =
+        try {
+          if (currentIterator eq null) {
```
Comment (nvollmar) [inline stream/src/main/scala/org/apache/pekko/stream/impl/fusing/IterableSource.scala:45]: couldn't we just initialize the variable with `elements.iterator` and avoid the if check/assignment here?

Comment (He-Pin) [reply stream/src/main/scala/org/apache/pekko/stream/impl/fusing/IterableSource.scala:45]: There are some cases where the `iterator` method call invoving some underling resources creation, that's why I defer the creation in the first place.

File: stream/src/main/scala/org/apache/pekko/stream/impl/fusing/IterableSource.scala [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+            decider(ex) match {
+              case Supervision.Stop   => failStage(ex)
+              case Supervision.Resume => tryPushNextOrComplete()
+              case Supervision.Restart =>
+                currentIterator = elements.iterator
+                tryPushNextOrComplete()
+            }
+        }
+
+      private def tryPushNextOrComplete(): Unit =
```
Comment (nvollmar) [inline stream/src/main/scala/org/apache/pekko/stream/impl/fusing/IterableSource.scala:60]: That would even be simpler:

```
      private def tryPushNextOrComplete(): Unit = {
         if (currentIterator.hasNext && isAvailable(out)) {
           push(out, currentIterator.next())
         }

         if (!currentIterator.hasNext) {
           completeStage()
         }
      }
```

Comment (He-Pin) [reply stream/src/main/scala/org/apache/pekko/stream/impl/fusing/IterableSource.scala:60]: But this will do another `currentIterator.hasNext` checking.

File: stream/src/main/scala/org/apache/pekko/stream/impl/fusing/IterableSource.scala (hunk 4 above)
Comment (He-Pin) [reply stream/src/main/scala/org/apache/pekko/stream/impl/fusing/IterableSource.scala:45]: IIRC the JOOQ does this, and our internal Lindorm Nosql db driver does the same too.
